TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java

```diff
@@ -467,11 +507,49 @@ private void writeEndOfSegmentMessage(Instant timestamp) throws IOException {
     }
 
     private void observationSizeSanityCheck(int minExpectedSpaceAfterObservation, int fieldNumber) throws IOException {
-        int actualRemainingSpace = getOrCreateCodedOutputStreamHolder().getOutputStreamSpaceLeft();
+        int actualRemainingSpace = currentOutputStreamWriteableSpaceLeft();
         if (actualRemainingSpace != -1 && (actualRemainingSpace < minExpectedSpaceAfterObservation || minExpectedSpaceAfterObservation < 0)) {
             log.warn("Writing a substream (capture type: {}) for Traffic Stream: {} left {} bytes in the CodedOutputStream but we calculated " +
                     "at least {} bytes remaining, this should be investigated", fieldNumber, connectionIdString + "." + (numFlushesSoFar + 1),
                 actualRemainingSpace, minExpectedSpaceAfterObservation);
         }
     }
+
+    private static class ByteOutputGatheringByteChannel implements GatheringByteChannel {
+        final ByteOutput byteOutput;
+
+        public ByteOutputGatheringByteChannel(ByteOutput byteOutput) {
+            this.byteOutput = byteOutput;
+        }
+
+        @Override
+        public int write(ByteBuffer src) throws IOException {
+            var bytesToWrite = src.remaining();
+            byteOutput.write(src);
+            return bytesToWrite - src.remaining();
+        }
+
+        @Override
+        public long write(ByteBuffer[] srcs, int offset, int length) throws IOException {
+            for (int i = offset; i < offset + length; i++) {
+                write(srcs[i]);
+            }
+            return length;
+        }
+
+        @Override
+        public long write(ByteBuffer[] srcs) throws IOException {
+            return write(srcs, 0, srcs.length);
+        }
+
+        @Override
+        public boolean isOpen() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void close() {
+            throw new UnsupportedOperationException();
```

Comment:
why would you penalize somebody for calling close()?  It's more future-proof to just leave this as a noop